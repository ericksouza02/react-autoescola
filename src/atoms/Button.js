import React from "react";
import PropTypes from "prop-types"
import styled, {css} from "styled-components";

const ButtonColors = {
    default: 'default',
    primary: 'primary',
    danger:  'danger',
}

const ButtonVariants = {
    outlined: 'outlined',
    link: 'link',
    default: 'default'
}

const getColorText = (props) => props => props.theme.colors.primary.text

const getMainColor = ({theme, color}) => {
    switch(color){
        case ButtonColors.primary:
            return theme.colors.primary.main;
        case ButtonColors.danger:
            return theme.colors.danger.main;
        default:
            return "#e0e0e0"
        }
}

const getDarkColor = ({theme, color}) => {
    switch(color){
        case ButtonColors.primary:
            return theme.colors.primary.dark;
        case ButtonColors.danger:
            return theme.colors.danger.dark;
        default:
            return "#e0e0e0"
        }
}





const Button = styled.button`    
    font-family: Arial, Helvetica, sans-serif;
    font-weight: 700;
    text-align: center;
    background-color: ${getMainColor};  
    border: 4px solid ${getMainColor};
    color: ${getColorText};
    letter-spacing: 1px;
    padding: 10px 30px;
    font-size: 20px;
    cursor: pointer;
    &:disabled {
        opacity: 0.6;
        cursor: not-allowed;
    }

    &:enabled {
        background-color: ${getDarkColor};
    }
`;

const ButtonOutlined = styled(Button)`    
    background-color: transparent;
    color: ${getMainColor};
    border: 2px solid;
`;

const ButtonLink = styled(Button)`    
    background-color: transparent;
    color: ${getMainColor};
    border-color: transparent;
    padding: 0;
`;


const ButtonWrapper = props => {
    switch(props.variant){
        case ButtonVariants.outlined:
            return <ButtonOutlined {...props}/>
        case ButtonVariants.link:
            return <ButtonLink {...props}/>
        default:
            return <Button {...props}/>
    }
}

ButtonWrapper.defaultProps = {
    type: 'button',
    color: 'default',
    variant: 'default',
    children: undefined
}

ButtonWrapper.propTypes = {
    type: PropTypes.string,
    color: PropTypes.oneOf(Object.values(ButtonColors)),
    variant: PropTypes.oneOf(Object.values(ButtonVariants)), 
    children: PropTypes.node,
}

export default ButtonWrapper;