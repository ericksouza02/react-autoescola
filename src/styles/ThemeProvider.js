import { ThemeProvider as StyledProvider } from "styled-components"

const theme = {
    colors:{
        primary: {
            main: '#F3D500',
            dark: '#E8BB00',
            light: '#FDED00',
            text: '#212121'
        },
        danger: {
            main: '#E80700',
            dark: '#D80300',
            light: '#F31F00',
            text: '#212121'
        }
    }
}

const ThemeProvider = ({children}) => (
    <StyledProvider theme={theme}>
        {children}
    </StyledProvider>
)

export default ThemeProvider;