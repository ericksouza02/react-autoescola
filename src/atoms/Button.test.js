import React from "react";
import { render, fireEvent } from '@testing-library/react'
import ThemeProvider from "../styles/ThemeProvider";

import Button from "./Button"

test('element children in the Button', () => {
    const { getByText } = render(
    <ThemeProvider>
        <Button>Clique Aqui</Button>
    </ThemeProvider>
)
    expect(getByText('Clique Aqui')).toBeInTheDocument()
});

test('event click on Button', () => {

    const handleCLick = jest.fn()

    const {getByRole} = render(
    <ThemeProvider>
        <Button onClick={handleCLick}/>
    </ThemeProvider>
    )

    fireEvent.click(getByRole('button')) // paramentros fireEvent.click(elemento, evento)

    expect(handleCLick).toBeCalled()
});

// Jeito de fazer com que o teste nÃ£o quebre
// Object.values(ButtonColors).map(item=>[item])

test.each([
   ['primary', 'outlined'],
   ['primary', 'link'],
   ['danger', 'outlined'],
   ['danger', 'link'],
   ['default', 'outlined'],
   ['default', 'link']
])('renders with %i color', (color, variant) => {

    const { asFragment } = render(
    <ThemeProvider>
        <Button color={color} variant={variant}/>
    </ThemeProvider>
    )

    expect(asFragment()).toMatchSnapshot()
});