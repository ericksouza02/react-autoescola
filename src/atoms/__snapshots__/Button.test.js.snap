// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders with NaN color 1`] = `
<DocumentFragment>
  <button
    class="sc-beyTiQ sc-guDMob eLbYwR flwiGj"
    color="primary"
    type="button"
    variant="outlined"
  />
</DocumentFragment>
`;

exports[`renders with NaN color 2`] = `
<DocumentFragment>
  <button
    class="sc-beyTiQ sc-dmyDmy eLbYwR hKJGHq"
    color="primary"
    type="button"
    variant="link"
  />
</DocumentFragment>
`;

exports[`renders with NaN color 3`] = `
<DocumentFragment>
  <button
    class="sc-beyTiQ sc-guDMob iXEzvn ksjnbt"
    color="danger"
    type="button"
    variant="outlined"
  />
</DocumentFragment>
`;

exports[`renders with NaN color 4`] = `
<DocumentFragment>
  <button
    class="sc-beyTiQ sc-dmyDmy iXEzvn fDCiuk"
    color="danger"
    type="button"
    variant="link"
  />
</DocumentFragment>
`;

exports[`renders with NaN color 5`] = `
<DocumentFragment>
  <button
    class="sc-beyTiQ sc-guDMob emNkql frdSKi"
    color="default"
    type="button"
    variant="outlined"
  />
</DocumentFragment>
`;

exports[`renders with NaN color 6`] = `
<DocumentFragment>
  <button
    class="sc-beyTiQ sc-dmyDmy emNkql nznlj"
    color="default"
    type="button"
    variant="link"
  />
</DocumentFragment>
`;
